src/docfx/build/toc/TableOfContentsMap.cs

```diff
@@ -0,0 +1,102 @@
+
+        private readonly ConcurrentDictionary<Document, Document> _nearestTocs = new ConcurrentDictionary<Document, Document>();
+
+        public TableOfContentsMap(IEnumerable<Document> tocs, Dictionary<Document, HashSet<Document>> documentToTocs)
+        {
+            _tocs = tocs ?? throw new ArgumentNullException(nameof(tocs));
+            _documentToTocs = documentToTocs ?? throw new ArgumentNullException(nameof(documentToTocs));
+        }
+
+        /// <summary>
+        /// Find the toc relative path to document
+        /// </summary>
+        /// <param name="file">Document</param>
+        /// <returns>The toc relative path</returns>
+        public string FindTocRelativePath(Document file)
+        {
+            var nearestToc = _nearestTocs.GetOrAdd(file, add => GetNearestToc(add));
+
+            return nearestToc != null ? PathUtility.NormalizeFile(PathUtility.GetRelativePathToFile(file.OutputPath, nearestToc.OutputPath)) : null;
+        }
+
+        /// <summary>
+        /// Return the nearest toc relative to the current file
+        /// "near" means less subdirectory count
+        /// when subdirectory counts are same, "near" means less parent directory count
+        /// e.g. "../../a/TOC.md" is nearer than "b/c/TOC.md"
+        /// </summary>
+        private Document GetNearestToc(Document file)
+        {
+            // fallback to all tocs if no toc files reference this file
+            var filteredTocFiles = _documentToTocs.TryGetValue(file, out var referencedTocFiles) ? referencedTocFiles : _tocs;
+
+            var nearstToc = (Document)null;
+            var nearestSubDirCount = 0;
+            var nearestParentDirCount = 0;
+            foreach (var toc in filteredTocFiles)
+            {
+                var relativePath = PathUtility.GetRelativePathToFile(toc.FilePath, file.FilePath);
+                var (subDirCount, parentDirCount) = GetDirectoryCount(relativePath);
+                if (nearstToc == null || Compare(nearestSubDirCount, nearestParentDirCount, subDirCount, parentDirCount) > 0)
+                {
+                    nearstToc = toc;
+                    nearestSubDirCount = subDirCount;
+                    nearestParentDirCount = parentDirCount;
+                }
+            }
+
+            return nearstToc;
+        }
+
+        private static int Compare(int xSubDirCount, int xParentDirCount, int ySubDirCount, int yParentDirCount)
+        {
+            if (xSubDirCount == ySubDirCount)
+            {
+                return xParentDirCount - yParentDirCount;
+            }
+            else
+            {
+                return xSubDirCount - ySubDirCount;
+            }
+        }
+
+        private static (int subDirectoryCount, int parentDirectoryCount) GetDirectoryCount(string relativePath)
+        {
+            relativePath = PathUtility.NormalizeFile(relativePath);
+            var relativePathParts = relativePath.Split('/');
```

Comment:
Can we avoid split string?

Reply:
more details: in `PathUtility.NormalizeFile`, we split path, then concat to one string, here we split it again.

Reply:
we can, just count ".."

---
In reply to: [184587292](https://github.com/dotnet/docfx/pull/2696#discussion_r184587292) [](ancestors = 184587292)

Reply:
i will change it later if necessary

---
In reply to: [184592262](https://github.com/dotnet/docfx/pull/2696#discussion_r184592262) [](ancestors = 184592262,184587292)

Reply:
please add `todo`

Reply:
OK

---
In reply to: [184836319](https://github.com/dotnet/docfx/pull/2696#discussion_r184836319) [](ancestors = 184836319)